modules/services/ollama.nix

```diff
@@ -0,0 +1,95 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.ollama;
+
+in {
+  options = {
+    services.ollama = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether to enable the Ollama Daemon.";
+      };
+
+      package = mkOption {
+        type = types.path;
+        default = pkgs.ollama;
+        description = "This option specifies the ollama package to use.";
+      };
+
+      host = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        example = "0.0.0.0";
+        description = ''
+          The host address which the ollama server HTTP interface listens to.
+        '';
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 11434;
+        example = 11111;
+        description = ''
+          Which port the ollama server listens to.
+        '';
+      };
+
+      home = lib.mkOption {
+        type = types.str;
+        default = "%S/ollama";
+        example = "/home/foo";
+        description = ''
+          The home directory that the ollama service is started in.
+        '';
+      };
+
+      models = mkOption {
+        type = types.str;
+        default = "%S/ollama/models";
```

Comment:
Need some help here. I didn't get why  in `nixpkgs`:

https://github.com/NixOS/nixpkgs/blob/be45e3445c7fd559f99f0581d807b1e32b381129/nixos/modules/services/misc/ollama.nix#L35

authors used string formatting specifier `%S` by default, without inserting anything there and just directly passing into environmental variable:

https://github.com/NixOS/nixpkgs/blob/be45e3445c7fd559f99f0581d807b1e32b381129/nixos/modules/services/misc/ollama.nix#L137

 In `ollama` [code](https://github.com/search?q=repo:ollama/ollama%20ModelsDir&type=code), I can't find any string formatting related to models destination path. Am I missing something?
 
> Also, actually, `ollama` uses `$HOME/.ollama/models`  by default:

https://github.com/ollama/ollama/blob/89c79bec8cf7a7b04f761fcc5306d2edf47a4164/envconfig/config.go#L294

Reply:
This is systemd‐specific syntax, documented in [`systemd.unit(5)`](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html). I believe it would ordinarily expand to `/var/lib`, but because of the `DynamicUser` sandboxing it acts slightly differently (per [`systemd.exec(5)`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html)):

> If` DynamicUser=` is used, the logic for `CacheDirectory=`, `LogsDirectory=` and `StateDirectory=` is slightly altered: the directories are created below `/var/cache/private`, `/var/log/private` and `/var/lib/private`, respectively, which are host directories made inaccessible to unprivileged users, which ensures that access to these directories cannot be gained through dynamic user ID recycling. Symbolic links are created to hide this difference in behaviour. Both from perspective of the host and from inside the unit, the relevant directories hence always appear directly below `/var/cache`, `/var/log` and `/var/lib`.

Since we don’t have the fancy `DynamicUser` sandboxing on macOS, you can just read this as `/var/lib`.

So basically we have a choice here. The thing that would best match NixOS is to define a user for this to run as (see `modules/services/gitlab-runner.nix`, `modules/services/hercules-ci-agent/default.nix`, `modules/services/ofborg/default.nix`, and `modules/services/buildkite-agents.nix` for examples), set the daemon to run as that user, and use paths under `/var/lib` here (although I don’t think that directory is commonly used on macOS). (I believe the current state of the PR would run Ollama as `root`, which is definitely not what we want.)

However, defining users on macOS is a bit annoying – they show up in System Settings and can potentially cause fuss on system upgrades unless you use a ~100 UID range that Apple keeps encroaching on and that Nix already puts 32 users in. So an alternative would be to define a `launchd.agent` instead and default this stuff to `$HOME/.ollama` or `$HOME/LIbrary/Application Support/ollama` or something. That would isolate Ollama less than the NixOS module does, and mean that every interactive user gets their own copy of the Ollama data, but I don’t know if that really matters.

Ideally we’d have a better story for defining daemon users, but for now the latter approach may be more practical.

Reply:
Okay, seems that I'll return `ollama.user.agents` back and use the same default as in [`ollama`](https://github.com/ollama/ollama/blob/main/envconfig/config.go#L293): `$HOME/.ollama/models`.